db/src/tx.rs

```diff
@@ -183,61 +188,204 @@ impl<'conn, 'a> Tx<'conn, 'a> {
-
                                 Either::Left(typed_value)
                             }
-                        };
+                        },
 
-                        Ok(Term::AddOrRetract(op, e, a, v))
-                    },
-                }
-            })
-            .collect::<Result<Vec<_>>>()
-            .map(|terms| (terms, temp_ids, lookup_refs))
+                        entmod::AtomOrLookupRefOrVectorOrMapNotation::LookupRef(lookup_ref) => {
+                            if attribute.value_type != ValueType::Ref {
+                                bail!(ErrorKind::NotYetImplemented(format!("Cannot resolve value lookup ref for attribute {} that is not :db/valueType :db.type/ref", a)))
+                            }
+
+                            Either::Right(LookupRefOrTempId::LookupRef(intern_lookup_ref(&mut lookup_refs, lookup_ref)?))
+                        },
+
+                        entmod::AtomOrLookupRefOrVectorOrMapNotation::Vector(vs) => {
+                            if !attribute.multival {
+                                bail!(ErrorKind::NotYetImplemented(format!("Cannot explode vector value for attribute {} that is not :db.cardinality :db.cardinality/many", a)));
+                            }
+
+                            for vv in vs {
+                                stack.push(Entity::AddOrRetract {
+                                    op: op.clone(),
+                                    e: e.clone(),
+                                    a: entmod::Entid::Entid(a),
+                                    v: vv,
+                                });
+                            }
+                            continue
+                        },
+
+                        entmod::AtomOrLookupRefOrVectorOrMapNotation::MapNotation(mut map_notation) => {
+                            // TODO: consider handling this at the tx-parser level.  That would be
+                            // more strict and expressive, but it would lead to splitting
+                            // AddOrRetract, which proliferates types and code, or only handling
+                            // nested maps rather than map values, like Datomic does.
+                            if op != OpType::Add {
+                                bail!(ErrorKind::NotYetImplemented(format!("Cannot explode nested map value in :db/retract for attribute {}", a)));
+                            }
+
+                            if attribute.value_type != ValueType::Ref {
+                                bail!(ErrorKind::NotYetImplemented(format!("Cannot explode nested map value for attribute {} that is not :db/valueType :db.type/ref", a)))
+                            }
+
+                            // :db/id is optional; if it's not given, we generate a special internal tempid
+                            // to use for upserting.  This tempid will not be reported in the TxReport.
```

Comment:
This was laziness.  The thing is that it's not splitting the `TempId` at parse time, it's splitting the `TempId` into external and internal at resolution time, which is a more complicated change.  (Possibly?)  However, it leads into associating partitions with tempids (ala #237) so I'll see how easy it is.